### ISSUE TITLE
Some problems with creating a scheduled task to run only once

### ISSUE THREAD

**Issue by mohmyo** [Koed00/django-q#475]
My goal is to create a scheduled task to run only **once** on a specified date, on the first try, it seemed good until my schedule got repeated.

So, I read about **repeats** As the docs pointed, the default value for **repeats** is -1 which means **always**, so I set it to 0 as it means **never** so, my schedule never repeats and only run once.

Here is an example of what I did

```python
schedule('app.tasks.auto_subscription_renew',
                     amount_cents, currency, instance.profile.id,name=schedule_name,repeats=0, schedule_type=Schedule.ONCE, next_run=end_date)
```
The problem is, when set to 0, the scheduler never run.
Then I find this hint about [Schedule.ONCE](https://django-q.readthedocs.io/en/latest/schedules.html#django_q.Schedule.ONCE)

So, I tried another value like -5, nothing changes, the same behaviour as -1

Also, what about the issue of multiple workers to catch the same task and run it twice? I had set it to 8 workers before but changed this to 1

Here is my conf, and for cache, I didn't set up one, the default in-memory cache in Django is being used for sure. 

```python
Q_CLUSTER = {
    'name': '_tasks',
    'workers': 1,
    'daemonize_workers': False,
    'retry': 3600,
    'compress': True,
    'save_limit': 0,
    'orm': 'temp',
    'bulk': 20
}
```

**Comment by mohmyo** [comment]
I have manually added a schedule through the admin panel too, first time with **repeats=0** and the schedule didn't run.
The second try updated the same schedule with repeats=1, this time schedule worked.

The same case for schedules in the past, when I change repeats to 1, schedules in the past started, nut with 0 it never runs.

**Comment by Koed00** [comment]
Hi Mahmoud. I'm currently on a trip, but as soon as I have some time on a
laptop I will see if I can re-create this issue.
Thanks for reporting it.

**Comment by mohmyo** [comment]
Enjoy your trip, have a nice time!

**Comment by mohmyo** [comment]
I have made a code walkthrough to find out why this is happening, so, I found [scheduler()](https://github.com/Koed00/django-q/blob/7b4c3571281c140b97ee4e5c00ff9a9345a383b6/django_q/cluster.py#L558) method which is according to the docstring in the code do the following: `Create a task from a schedule at the scheduled time and schedules next run`, and It got called once a minute (or so) by [Sentinel.guard()](https://github.com/Koed00/django-q/blob/7b4c3571281c140b97ee4e5c00ff9a9345a383b6/django_q/cluster.py#L273).

The first issue is that schedules with repeats=0 got [excluded ](https://github.com/Koed00/django-q/blob/7b4c3571281c140b97ee4e5c00ff9a9345a383b6/django_q/cluster.py#L569) and I didn't find something else anywhere that pushes schedules with repates=0 to the task queuing later.

[The second one is](https://github.com/Koed00/django-q/blob/7b4c3571281c140b97ee4e5c00ff9a9345a383b6/django_q/cluster.py#L646), is never run as this way. A schedule with a negative or positive number or repeats is always there and get a repeated execution. I think this happens because pushing to task queue is [done before](https://github.com/Koed00/django-q/blob/7b4c3571281c140b97ee4e5c00ff9a9345a383b6/django_q/cluster.py#L631) the mentioned code block.

I will try to do manual patching on my current working space and see if fixing these points will fix the issue or not.

**Comment by mohmyo** [comment]
Ok, all of the previous points have failed miserably, removed the exclude(repeats=0) and rearranged schedule.ONCE block to come before pushing a scheduled task and nothing changes, the schedule is repeated indefinitely.

**Comment by Koed00** [comment]
I did some testing myself in one of my production enviroments, cause I haven't used this feature in a while and these were the results:

**Schedule Once, repeats -1** : Ran one time, scheduled tasks got deleted, result was in the succesful tasks table.
**Schedule Once, repeats 1**:  Ran one time, scheduled tasks remains with repeats 0, result in the succesful tasks table. No repeats
**Schedule Once, repeats 0:**  Does not run 

This is what I expect to happen.  We added the repeats +1 variant for people who want to keep the schedule and reschedule it at some point, instead of creating a new one. This is mostly for those who use the admin.
The 0 repeats state works the same for all schedule types and will pause the schedule until it the value is either positive or negative.

Unless I don't fully understand your particular situation, you should create your once schedules with repeats 1.
Let me know if you encounter any weird behaviour.

**Comment by mohmyo** [comment]
Thanks for replying, then before trying again, is my Q_CLUSTER conf valid? I thought after all, maybe I'm doing something wrong, that caused this behaviour.

**Comment by mohmyo** [comment]
Ok, I have tested now with the mentioned scenarios in mind, at least now I can say something new.

The problem turned out to be that the number of repeats is **not changing at all**, that's why the task got repeated no matter what is the value of repeats because it is still the same, so the schedule is got picked up by cluster every time.

I have figured this out when added a schedule at the admin panel with repates=1, as mentioned, repeats should be 0 then, but after the first time, it got executed it still 1 and all along.

I'm suspecting my DB as I'm working on WSL for the first time because django-q doesn't work on Windows 10, but it's weird since everything else is working fine and got saved to the DB, so why it will only happen with updating repeats.

**Comment by mohmyo** [comment]
OK, as a final step to make sure that this doesn't relate to developing in WSL, I have tested the scheduling behaviour on our staging server -which runs **ubuntu18 with python 3.7.9** and is working just fine- **to get the same exact behaviour of schedules gets repeated as I experienced before**

I'm sure now that nothing wrong with the environment, so it must be something with  `Q_CLUSTER ` conf or something else hidden.

**Comment by mohmyo** [comment]
@Koed00 I can now say where the problem is!

First, created a new Django project to make sure and start testing, only added django_q to this project, used ORM as a broker with the **`default`** database. schedules worked as expected this time and repeats value got changed too.

**But here is the interesting part**, when created a second database and updated **`Q_CLUSTER`** to use it instead of default, the exact weird behaviour I had for the last week is there! this means that using a different database than the default is not working as expected! then tried using default database in my original project instead of a different one and it worked.

I made sure that migrations were applied to both databases- default and the other DB- , even with migrations that aren't needed for the second one, just to make sure this is not a migration issue, -anyway no exceptions were thrown anytime so tables are there but just to make sure-

I'm using django-q 1.3.3, the new project to test was Django 3.1 with SQLite for DB, and my existing project is Django 2.2.16 with PostgreSQL 12.3

**Comment by Koed00** [comment]
Ok I understand the problem now; one of the contributros added the option to have the queue on a different database, but it was never intended to run everything, including the admin stuff, on a separate db connection.
There is another issue on this somehwere, but I can't quite find it right now.

We would have to the option to run everything on a different db for your setup to completely work. Which I'm all for.
A quick workaround could be to add the schedule object (and other objects) to a database router and let django know where to find them by default.

**Comment by mohmyo** [comment]
Thank you so much, I have a better understanding for this now, but I feel there are plenty of points that should be added to the documentation to make it more clear and stop confusion because I only followed it and got to the current situation.

Now, I will compare between running on the `default` for now and switch to a separate later when it's complete, or do the workaround you mentioned, it depends on the scale I'm expecting, so I should reconsider this again.

Thanks again for your help and your effort to make this package great!

**Comment by adamduren** [comment]
@Koed00 I am also experiencing this issue with version `1.3.4` and I don't quite understand the workaround you described.

In our setup we use the ORM setup and only have one database defined. `async_task` works as expected and `schedule` works in the `hourly` and `daily` cases but not for `once`.

**Comment by adamduren** [comment]
I was able to resolve my issue. Not sure if it is related but I detailed my solution in the comment below

https://github.com/Koed00/django-q/issues/400#issuecomment-719037271

**Comment by jpaav** [comment]
I am also experiencing this issue with the ORM broker on version `1.3.3` and I can confirm that switching to a single database, instead of using a separate one for the broker, fixed the behavior.

**Comment by plungerman** [comment]
i have the same problem using 1.3.5, python 3.6, django 2.2.20, on ubuntu 18.04. i am attempting to schedule something to run once in the future. here it is just 1 minute in the future, as example:

schedule(
    'path.to.func',
    args,
    schedule_type=Schedule.ONCE,
    next_run=timezone.now() + timedelta(minutes=1),
    repeats=-1,
)

if repeats is 0, then it never runs. when repeats is -1 or any negative number it continues to repeat indefinitely. when repeats=1 the schedule runs twice, and the Schedule objects remains with repeats=0 and another "instance" in the "Queued tasks" list in django admin, and that continues to execute every minute until i delete the instance from the "Queued tasks".

any thoughts on how to set up a schedule to execute once in the future and then never again?

**Comment by RajuSingh-cmyk** [comment]
i want to run djangoq-scheduler on every weekend but having two issues with it-:

 

i) creating two entries in django_q_scheduler models as soon as i run development server and qcluster

ii) scheduler starts executing immediately as soon as entry is created in the model, instead of executing on every weekend.

 

Here is the code used -:

 

# Django-Q Cluster

Q_CLUSTER = {

    'name': 'DjangQORM',

    'max_attempts': 1,

    'retry': 14400,

    'orm': 'default'

}

 

 

Schedule.objects.create(func='func()',

                        schedule_type=Schedule.CRON,

                        cron='30 18 * * 6'

                        )

 

 

Version Used -

python-> 3.6.8

django-q -> 1.3.4

 

Database- postgres

OS - Linux

**Comment by plungerman** [comment]
> i have the same problem using 1.3.5, python 3.8, django 2.2.20, on ubuntu 20.04. i am attempting to schedule something to run once in the future. here it is just 1 minute in the future, as example:
> 
> ```
> schedule(
>     'path.to.func',
>     args,
>     schedule_type=Schedule.ONCE,
>     next_run=timezone.now() + timedelta(minutes=1),
>     repeats=-1,
> )
> ```
> 
> if repeats is 0, then it never runs. when repeats is -1 or any negative number it continues to repeat indefinitely. when repeats=1 the schedule runs one, and the Schedule objects remains with repeats=0 and another "instance" in the "Queued tasks" list in django admin, and that continues to execute every minute until i delete the instance from the "Queued tasks".
> 
> any thoughts on how to set up a schedule to execute once in the future and then never again?

i just wanted to provide an update on this issue.

i installed redis-server to use instead of the django ORM and that solved the problem. one thing to note, however, is that just removing the orm parameter from the Q_CLUSTER dictionary did not solve the problem. there was an entry in the django_q_ormq table that i had to delete as well. once that was done, the repeats parameter for Schedule() worked as documented.
